### PR TITLE
Added event replicator and unit tests

### PR DESCRIPTION
**Pull request by li4song4**
## What is this PR for?
Added code for event replicator that will replicate event data from CDF and store in a Lakehouse table, with the table path and name configurable.

Also added unit tests for the event replicator.

## What does the code do?
- The code for the event replicator is in "event.py".
- Adjusted config to incorporate configurations for event replicator including batch size and abfss table path
- Added unit tests for the replicator in "tests/unit/test_event.py"

## How do I verify the changes made?
- Copy the ".env.example" to ".env" and populate proper values for all environment variables. The environment variable that's specifically for event replicator is `EVENT_TABLE_NAME`
- The default batch size in "example_config.yaml" is set to 5. Change the batch size if needed.
- [optional] In "__main__.py", comment out thread creation for other replicators to focus resources on the event replicator
- Run `poetry run cdf_fabric_replicator example_config.yaml` to run the event replicator

image

image

- Run `poetry run pytest tests/unit/test_event.py` to verify that all unit tests pass

image

### CODE REVIEW

File: cdf_fabric_replicator/event.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            name="cdf_fabric_replicator_data_modeling",
+            description="CDF Fabric Replicator",
+            config_class=Config,
+            metrics=metrics,
+            use_default_state_store=False,
+            version=__version__,
+            cancellation_token=stop_event,
+        )
+        self.metrics: Metrics
+        # self.stop_event = stop_event
```
Comment (kenny-chalupa) [inline cdf_fabric_replicator/event.py:30]: does this comment still need to be here?

File: cdf_fabric_replicator/event.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+from cdf_fabric_replicator import __version__
+from cdf_fabric_replicator.config import Config
+from cdf_fabric_replicator.metrics import Metrics
+from cognite.client.data_classes import EventList
+
+class EventsReplicator(Extractor):
+    def __init__(self, metrics: Metrics, stop_event: CancellationToken) -> None:
+        super().__init__(
+            name="cdf_fabric_replicator_data_modeling",
```
Comment (kenny-chalupa) [inline cdf_fabric_replicator/event.py:21]: we should change this name to the events

File: cdf_fabric_replicator/event.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        while True:  # not self.stop_event.is_set():
+            start_time = time.time()  # Get the current time in seconds
+
+            self.process_events()
+            end_time = time.time()  # Get the time after function execution
+            elapsed_time = end_time - start_time
+            sleep_time = max(
+                self.config.extractor.poll_time - elapsed_time, 0
+            )  # 900s = 15min
```
Comment (kenny-chalupa) [inline cdf_fabric_replicator/event.py:52]: do we still need this comment?

File: tests/unit/test_event.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    event_replicator.client = event_replicator.config.cognite.get_cognite_client("test_event_replicator")
+    event_replicator.state_store = LocalStateStore(
+        event_replicator.config.extractor.state_store.local.path
+    )
+
+    return event_replicator
+
+@pytest.fixture
+def early_created_time():
+    return 1710503304020
```
Comment (kenny-chalupa) [inline tests/unit/test_event.py:22]: Can we use this as a constant instead EARLY_CREATED_TIME = 1710503304020

File: tests/unit/test_event.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+@pytest.fixture
+def late_created_time():
+    return 1710506651232
```
Comment (kenny-chalupa) [inline tests/unit/test_event.py:26]: Same here to try to use a constant instead

File: tests/unit/test_event.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    mock_write_events_to_lakehouse_tables.assert_called_with(
+        event_data_camel_case, 
+        event_replicator.config.event.lakehouse_abfss_path_events
+    )
+    mock_set_event_state.assert_called_with(
+        event_replicator.event_state_key, 
+        event_data_camel_case[-1]["createdTime"]
+    )
+
+def test_write_late_events_to_fabric(event_replicator, early_created_time, event_data_camel_case, mock_get_late_events, mock_set_event_state, mock_write_events_to_lakehouse_tables, mocker):
```
Comment (kenny-chalupa) [inline tests/unit/test_event.py:110]: here is something to consider, as we are adding more and more fixture references to these tests. It may be beneficial to add type referencing to help keep track of what to expect from all these different fixtures:

```python
def test_write_late_events_to_fabric(event_replicator: EventsReplicator, early_created_time: Literal[1710503304020], event_data_camel_case: list[dict], mock_get_late_events: Any, mock_set_event_state: Any, mock_write_events_to_lakehouse_tables: Any, mocker):
```

File: cdf_fabric_replicator/event.py (hunk 1 above)
Comment (li4song4) [reply cdf_fabric_replicator/event.py]: removed

File: cdf_fabric_replicator/event.py (hunk 2 above)
Comment (li4song4) [reply cdf_fabric_replicator/event.py]: changed

File: cdf_fabric_replicator/event.py (hunk 3 above)
Comment (li4song4) [reply cdf_fabric_replicator/event.py]: Removed

File: tests/unit/test_event.py (hunk 4 above)
Comment (li4song4) [reply tests/unit/test_event.py]: Originally thought it would make it clearer to be referred among other fixtures and test methods, but doesn't seem worth the complication. Used a constant instead to simplify the code.

File: tests/unit/test_event.py (hunk 5 above)
Comment (li4song4) [reply tests/unit/test_event.py]: Changed to a constant

File: tests/unit/test_event.py (hunk 6 above)
Comment (li4song4) [reply tests/unit/test_event.py]: Added type indicators.

**Comment by kenny-chalupa** [comment]
verified event were properly replicated in the lakehouse table:

image

as well as the unit tests verification:

image

File: cdf_fabric_replicator/event.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        super().__init__(
+            name="cdf_fabric_replicator_events",
+            description="CDF Fabric Replicator",
+            config_class=Config,
+            metrics=metrics,
+            use_default_state_store=False,
+            version=__version__,
+            cancellation_token=stop_event,
+        )
+        self.metrics: Metrics
```
Comment (nicholasperkins1992) [inline cdf_fabric_replicator/event.py:29]: this is defined but does not seem to be used anywhere

File: cdf_fabric_replicator/event.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        self.endpoint_source_map: Dict[str, Any] = {}
+        self.errors: List[str] = []
```
Comment (nicholasperkins1992) [inline cdf_fabric_replicator/event.py:31]: this is defined but does not seem to be used anywhere

File: tests/unit/test_event.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    mock_get_events.assert_called_with(event_replicator.config.event.batch_size, 0)
+    mock_write_events_to_lakehouse_tables.assert_called_once()
+    mock_write_events_to_lakehouse_tables.assert_called_with(
+        event_data_camel_case, 
+        event_replicator.config.event.lakehouse_abfss_path_events
+    )
+    mock_set_event_state.assert_called_with(
+        event_replicator.event_state_key, 
+        event_data_camel_case[-1]["createdTime"]
+    )
```
Comment (nicholasperkins1992) [inline tests/unit/test_event.py:110]: These test might be easier to read if has Arrange Act Assert sections in each one

File: tests/unit/test_event.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    return components[0] + ''.join(x.title() for x in components[1:])
+
+def convert_dict_keys_to_camel_case(snake_case_dict):
+    return {snake_to_camel(k): v for k, v in snake_case_dict.items()}
+
+@pytest.fixture
+def event_data_camel_case(event_data):
+    return [convert_dict_keys_to_camel_case(data) for data in event_data]
+
+def test_write_all_events_to_fabric(
```
Comment (nicholasperkins1992) [inline tests/unit/test_event.py:92]: Generally for unit tests you should be testing a specific method, this name seems like you are testing writing to Fabric which i think would be a system or integration test

Comment (nicholasperkins1992) [reply tests/unit/test_event.py:92]: You may want to name it something like test_Process_Events_Should_NotThrow

File: tests/unit/test_event.py (hunk 9 above)
Comment (li4song4) [reply tests/unit/test_event.py:110]: I'll add the sections for these unit tests since they are applicable here. However, please note that sometimes Pytest may not have the "arrange" section as they sometimes are taken care of by fixtures.

File: tests/unit/test_event.py (hunk 10 above)
Comment (li4song4) [reply tests/unit/test_event.py]: Good advice. Change the name to relate to the method that's tested.

File: cdf_fabric_replicator/event.py (hunk 8 above)
Comment (li4song4) [reply cdf_fabric_replicator/event.py]: Removed.

File: cdf_fabric_replicator/event.py (hunk 7 above)
Comment (li4song4) [reply cdf_fabric_replicator/event.py]: Removed

File: tests/unit/test_event.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+@pytest.fixture
+def mock_set_event_state(mocker):
+    return mocker.patch("cdf_fabric_replicator.event.EventsReplicator.set_event_state", return_value=None)
+
+@pytest.fixture
+def mock_write_events_to_lakehouse_tables(mocker):
+    return mocker.patch("cdf_fabric_replicator.event.EventsReplicator.write_events_to_lakehouse_tables", return_value=None)
+
+def snake_to_camel(snake_str):
```
Comment (wbevans) [inline tests/unit/test_event.py:81]: Should probably add type hints to method.

File: tests/unit/test_event.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            'asset_ids': [5534771458240925], 
+            'id': 5875589750425323, 
+            'last_updated_time': 1710506651232, 
+            'created_time': LATE_CREATED_TIME
+        }
+    ]
+
+@pytest.fixture
+def mock_get_events(mocker, event_data):
+    return_value = iter([EventList([Event(**data) for data in event_data])])
```
Comment (wbevans) [inline tests/unit/test_event.py:59]: A bit hard to follow what the logic is doing here with everything inline. Might be easier to follow if you break it out into a few intermediate steps.

Comment (li4song4) [reply tests/unit/test_event.py]: Modified.

File: tests/unit/test_event.py (hunk 11 above)
Comment (li4song4) [reply tests/unit/test_event.py]: Added type hints to both methods.

**Review by muradsater — APPROVED** [review]
